tests/Api/Monetization/Denormalizer/ReportCriteriaDenormalizerTest.php

```diff
@@ -0,0 +1,113 @@
+<?php
+
+/*
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Apigee\Edge\Tests\Api\Monetization\Denormalizer;
+
+use Apigee\Edge\Api\Monetization\Denormalizer\ReportCriteriaDenormalizer;
+use Apigee\Edge\Api\Monetization\Entity\EntityInterface;
+use Apigee\Edge\Api\Monetization\Entity\ReportDefinitionInterface;
+use Apigee\Edge\Api\Monetization\Structure\Reports\Criteria\AbstractCriteria;
+use Apigee\Edge\Api\Monetization\Structure\Reports\Criteria\BillingReportCriteria;
+use Apigee\Edge\Api\Monetization\Structure\Reports\Criteria\PrepaidBalanceReportCriteria;
+use Apigee\Edge\Api\Monetization\Structure\Reports\Criteria\RevenueReportCriteria;
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\Serializer\Normalizer\DateTimeNormalizer;
+use Symfony\Component\Serializer\Serializer;
+
+class ReportCriteriaDenormalizerTest extends TestCase
+{
+    /** @var \Apigee\Edge\Api\Monetization\Denormalizer\ReportCriteriaDenormalizer */
+    protected static $denormalizer;
+
+    /**
+     * @inheritDoc
+     */
+    public static function setUpBeforeClass(): void
+    {
+        parent::setUpBeforeClass();
+        static::$denormalizer = new ReportCriteriaDenormalizer();
+        static::$denormalizer->setSerializer(new Serializer([new DateTimeNormalizer([DateTimeNormalizer::FORMAT_KEY => EntityInterface::DATE_FORMAT])]));
+    }
+
+    /**
+     * @expectedException \Error
+     */
+    public function testDenormalizeWithAbtractClassNoContext(): void
+    {
+        static::$denormalizer->denormalize((object) [], AbstractCriteria::class, 'json');
+    }
+
+    /**
+     * @expectedException \Apigee\Edge\Exception\InvalidArgumentException
```

Comment:
There is this article about expected exceptions: https://thephp.cc/news/2016/02/questioning-phpunit-best-practices 

The logic layed out there seems quite sensible, even it matters less here, because there is just a single line in this test.

Reply:
@dawehner Thanks for this article, it was an interesting reading even it is 3 years old.
In my next project, I'll not use the annotation based exception testing, because I do not like that either :) But as you wrote, here it does not really matter which one I use and for consistency reasons, I'd keep the current implementation.